### PR TITLE
Analysis tools for camera calibration / hand-eye calibration

### PR DESCRIPTION
**Pull request by s-trinh**
Some tools to try to analyze the camera calibration results.

---

Running `./calibrate-camera default-chessboard.cfg`:

- this shows the locations of the calibration patterns in the image

image

- this shows the current reprojection error, the extracted points and the projected points using the estimated cMo and camera parameters

image

- left image: draw lines starting from the first and last points using points (`findChessboardCorners`). This should allow seeing how distorted is the image.
- right image: image is undistorted, chessboard points are extracted and line fitting is computed. Draw lines from 2D line equations estimated on points extracted from undistorted image (to check that the line fitting is correct)

image

- output of the console is RMS error for line fitting for points extracted from distorted image and after using `vpPixelMeterConversion::convertPoint`:
```
Line fitting error on distorted points: 0.2630394859 ; on undistorted points: 0.1779744522
Line fitting error on distorted points: 0.2397191393 ; on undistorted points: 0.1883295563
Line fitting error on distorted points: 0.1886317017 ; on undistorted points: 0.1713784167
Line fitting error on distorted points: 0.1453686602 ; on undistorted points: 0.1574874039
Line fitting error on distorted points: 0.08931020024 ; on undistorted points: 0.1317065326
Line fitting error on distorted points: 0.03638023684 ; on undistorted points: 0.09802067829
```

- output of the console is RMS error for line fitting for points extracted from undistorted image:
```
Undistorted image, line fitting error: 0.1631801528
Undistorted image, line fitting error: 0.198887235
Undistorted image, line fitting error: 0.1757022415
Undistorted image, line fitting error: 0.1563067239
Undistorted image, line fitting error: 0.1330949259
Undistorted image, line fitting error: 0.07752279553
```

---

Running `python camera_calibration_show_extrinsics.py --calibration camera.xml --scale_focal 20`:

image

- this shows the camera poses, the `camera.xml` must be generated using code from this PR to save the camera extrinsics

---

Running `python hand_eye_calibration_show_extrinsics.py --ndata 8 --eMc_yaml eMc.yaml`:

image

- this shows the camera poses used for hand-eye calibration and the end-effector poses using the estimated `eMc`

---

The files `camera_calibration_show_extrinsics.py` and `hand_eye_calibration_show_extrinsics.py` are based on this [OpenCV code](https://github.com/opencv/opencv/blob/4.3.0/samples/python/camera_calibration_show_extrinsics.py). I have put the BSD-3 license to not have GPL2 constraint and be able to use these scripts without limitations.

### CODE REVIEW

**Comment by s-trinh** [comment]
To test the image undistortion, there are some images [here](https://github.com/opencv/opencv/blob/4.3.0/samples/data/left01.jpg). Some files need to be renamed since there is no `left10.jpg`.

---

About the calibration process, I think guidelines should be:

- use a rigid planar calibration board, a printed sheet is not ideal
- focused images, adjust the camera parameters to have good illumination

Looks like there is no need to use lots of image, but rather good poses:
- in `rc_visard` camera calibration tool, it uses 5 poses
- here [some comments](https://github.com/xoox/calibrel/issues/1#issuecomment-419279616) about the calibration process (avoid perpendicular images)
- [DLR CalDe](https://www.dlr.de/rm/en/desktopdefault.aspx/tabid-3925/6084_read-9196/) recommendations

Most likely the 3 mentioned links use the DLR method. Not sure if these recommentations are still relevant with the underlying method in ViSP.

---

This copies also the [Matlab calibration tutorial](https://www.mathworks.com/help/vision/examples/evaluating-the-accuracy-of-single-camera-calibration.html).
I think additional recommendations should be:
- camera calibration is a trial and error process
- for instance, is it a good idea to do a first run and remove the images with high reprojection errors?
- the first run should allow removing blurry images, or images where corners are not accurately extracted

**Comment by s-trinh** [comment]
By the way, I have changed a little the "tool".

Now it draws lines from extracted corners for raw images and undistorted images. It uses first and last coordinates for each calibration row in both cases. Before it draws lines from line fitting.

Only in the console there are line fitting errors for:
- corners extracted from raw images, to try to see how strong the distortion is
- corners extracted from undistorted images, to see if the images are correctly undistorted
- corners extracted from raw images, but converted using something like `undistortPoint`

The 2 latter cases are different since undistorting the images and undistorting the point coordinates are a little bit different. Because there are `kud` and `kdu` parameters.
To undistort the images, we use one of the two parameters to compute the distorted pixel coordinates and interpolate its value, while to undistort the point coordinates it is the other parameters.

**Comment by fspindle** [comment]
@s-trinh Ready to merge on my side

File: doc/tutorial/calibration/tutorial-calibration-intrinsic.doc [hunk 1]
```diff
@@ -204,84 +281,304 @@ Square_Size: 0.034
...
+
+\subsubsection calibration_tools_repro_error Reprojection error
+
+Reprojection error could be seen in the next image. It shows the current reprojection error, the extracted points and the projected points using the estimated `cMo` camera pose and camera parameters:
+\image html img-calib-reprojection-error.jpg
+
+On the console, the global reprojection error is also given:
+\verbatim
+Calibration without distortion in progress on 5 images...
+Global reprojection error: 1.552697985
```
Comment (s-trinh) [inline doc/tutorial/calibration/tutorial-calibration-intrinsic.doc:384]:
```suggestion
Calibration without distortion in progress on 13 images...
Camera parameters for perspective projection without distortion:
  px = 557.2519329   py = 561.1776205
  u0 = 360.0754485   v0 = 235.5240362

Image chessboard-01.jpg reprojection error: 1.230158891
Image chessboard-02.jpg reprojection error: 1.447042293
Image chessboard-03.jpg reprojection error: 2.078525011
Image chessboard-04.jpg reprojection error: 1.553888654
Image chessboard-05.jpg reprojection error: 1.697849369
Image chessboard-06.jpg reprojection error: 2.284910081
Image chessboard-07.jpg reprojection error: 1.387893707
Image chessboard-08.jpg reprojection error: 1.664988951
Image chessboard-09.jpg reprojection error: 0.9447032921
Image chessboard-10.jpg reprojection error: 1.258540894
Image chessboard-11.jpg reprojection error: 1.845113925
Image chessboard-12.jpg reprojection error: 0.8621449347
Image chessboard-13.jpg reprojection error: 1.254269691

Global reprojection error: 1.552697985
```

File: doc/tutorial/calibration/tutorial-calibration-intrinsic.doc [hunk 2]
```diff
@@ -204,84 +281,304 @@ Square_Size: 0.034
...
+
+Calibration with distortion in progress on 5 images...
+Global reprojection error: 0.4338765096
```
Comment (s-trinh) [inline doc/tutorial/calibration/tutorial-calibration-intrinsic.doc:387]:
```suggestion
Calibration with distortion in progress on 13 images...
Camera parameters for perspective projection with distortion:
  px = 536.982404    py = 537.4196388
  u0 = 344.0958214   v0 = 235.1792108
  kud = -0.2672672093
  kdu = 0.3081092949

Image chessboard-01.jpg reprojection error: 0.2395321578
Image chessboard-02.jpg reprojection error: 1.238545486
Image chessboard-03.jpg reprojection error: 0.3207755448
Image chessboard-04.jpg reprojection error: 0.2601958886
Image chessboard-05.jpg reprojection error: 0.2338664574
Image chessboard-06.jpg reprojection error: 0.3297641256
Image chessboard-07.jpg reprojection error: 0.2475833215
Image chessboard-08.jpg reprojection error: 0.2602460339
Image chessboard-09.jpg reprojection error: 0.3159699755
Image chessboard-10.jpg reprojection error: 0.198531964
Image chessboard-11.jpg reprojection error: 0.2287278629
Image chessboard-12.jpg reprojection error: 0.406199404
Image chessboard-13.jpg reprojection error: 0.1950310138

Global reprojection error: 0.4338765096
```

File: doc/tutorial/calibration/tutorial-calibration-intrinsic.doc [hunk 3]
```diff
@@ -204,84 +281,304 @@ Square_Size: 0.034
...
+\endverbatim
+
+A good calibration is obtained when extracted and reprojected points are very close each other and when global reprojection error is less than 1.0. Smaller is the reprojection error, better is the calibration.
+
+\subsubsection calibration_tools_line_fit Line fitting
+
+To get an idea on how much there is distortion, `calibrate-camera` binary generates also the following images:
+- left image: draw lines starting from the first and last points using points findChessboardCorners() function from OpenCV. This should allow seeing how distorted is the image.
+- right image: image is undistorted, chessboard points are extracted and line fitting is computed. Here lines are drawn from 2D line equations estimated on points extracted from undistorted image (to check that the line fitting is correct)
```
Comment (s-trinh) [inline doc/tutorial/calibration/tutorial-calibration-intrinsic.doc:396]:
```suggestion
- left image: draw lines starting from the first and last points using points `findChessboardCorners()` function from OpenCV. This should allow seeing how distorted is the image.
- right image: image is undistorted, chessboard points are extracted and lines starting from the first and last points are drawn
```

File: doc/tutorial/calibration/tutorial-calibration-intrinsic.doc [hunk 4]
```diff
@@ -204,84 +281,304 @@ Square_Size: 0.034
...
+
+\image html img-calib-line-fitting.jpg
+
+On the console, RMS error for line fitting for points extracted from distorted image and after using vpPixelMeterConversion::convertPoint() is given:
```
Comment (s-trinh) [inline doc/tutorial/calibration/tutorial-calibration-intrinsic.doc:400]:
```suggestion
On the console, mean distance errors for line fitting for points extracted from distorted image and after using vpPixelMeterConversion::convertPoint() are given:
```

File: doc/tutorial/calibration/tutorial-calibration-intrinsic.doc [hunk 5]
```diff
@@ -204,84 +281,304 @@ Square_Size: 0.034
...
+This tool computes the line fitting error (mean distance error) on image points extracted from the raw distorted image.
+chessboard-03.jpg line 1 fitting error on distorted points: 1.524931112 ; on undistorted points: 0.2997221937
+chessboard-03.jpg line 2 fitting error on distorted points: 1.161437677 ; on undistorted points: 0.1608990656
+chessboard-03.jpg line 3 fitting error on distorted points: 0.7274089743 ; on undistorted points: 0.1089682074
+chessboard-03.jpg line 4 fitting error on distorted points: 0.2495991137 ; on undistorted points: 0.0648610129
+chessboard-03.jpg line 5 fitting error on distorted points: 0.2932864898 ; on undistorted points: 0.07897041729
+chessboard-03.jpg line 6 fitting error on distorted points: 0.9474403828 ; on undistorted points: 0.03407551552
+\endverbatim
+
+RMS error for line fitting for points extracted from undistorted image is also printed on the console:
```
Comment (s-trinh) [inline doc/tutorial/calibration/tutorial-calibration-intrinsic.doc:411]:
```suggestion
Mean distance errors for line fitting for points extracted from undistorted image are also printed on the console:
```

**Comment by s-trinh** [comment]
@fspindle 
Some suggestions. Ready to merge also.

**Comment by fspindle** [comment]
@s-trinh Thanks.

**Pull request merged**